tornado-drivers/drivers-common/src/main/java/uk/ac/manchester/tornado/drivers/common/TornadoBufferProvider.java

```diff
@@ -44,8 +46,8 @@
 public abstract class TornadoBufferProvider {
 
     protected final TornadoDeviceContext deviceContext;
-    protected final List<BufferContainer> freeBuffers;
-    protected final List<BufferContainer> usedBuffers;
+    protected final HashMap<Access, ArrayList<BufferContainer>> freeBuffers;
```

Comment:
Use `Map` and allocate a ConcurrentHashMap. I think there is the possibility to run this code from multiple threads.